src/ts/index.ts

```diff
@@ -0,0 +1,291 @@
+   * For fill or kill sell orders, this amount represents the exact sell amount
+   * that will be executed in the trade. For fill or kill buy orders, this
+   * amount represents the maximum sell amount that can be executed. For partial
+   * fill orders, this represents a component of the limit price fraction.
+   */
+  sellAmount: BigNumberish;
+  /**
+   * The order buy amount.
+   *
+   * For fill or kill sell orders, this amount represents the minimum buy amount
+   * that can be executed in the trade. For fill or kill buy orders, this amount
+   * represents the exact buy amount that will be executed. For partial fill
+   * orders, this represents a component of the limit price fraction.
+   */
+  buyAmount: BigNumberish;
+  /**
+   * The timestamp this order is valid until
+   */
+  validTo: number | Date;
+  /**
+   * A nonce to ensure orders aren't replayable.
+   *
+   * Specifying a value of {@link REPLAYABLE_NONCE} indicates that this order
+   * can be replayed in any batch. This can be used for setting up passive
+   * trading strategies where orders are valid forever.
+   */
+  nonce: number;
+  /**
+   * Additional fee to give to the protocol.
+   *
+   * This tip is used to offset solution submission gas costs so orders that
+   * aren't economically viable (i.e. small orders that do not generate enough
+   * fees) still get executed.
+   */
+  tip: BigNumberish;
+  /**
+   * The order kind.
+   */
+  kind: OrderKind;
+  /**
+   * Specifies whether or not the order is partially fillable.
+   */
+  partiallyFillable: boolean;
+}
+
+/**
+ * Gnosis Protocol v2 order flags.
+ */
+export type OrderFlags = Pick<Order, "kind" | "partiallyFillable">;
+
+/**
+ * Order kind.
+ */
+export const enum OrderKind {
+  /**
+   * A sell order.
+   */
+  SELL,
+  /**
+   * A buy order.
+   */
+  BUY,
+}
+
+/**
+ * Reserved nonce value used to indicate that an order can be replayed in any
+ * batch.
+ */
+export const REPLAYABLE_NONCE = 0;
+
+function timestamp(time: number | Date): number {
+  return typeof time === "number" ? time : ~~(time.getTime() / 1000);
+}
+
+function encodeOrderFlags(flags: OrderFlags): number {
+  const kind = flags.kind === OrderKind.SELL ? 0x00 : 0x01;
+  const partiallyFillable = flags.partiallyFillable ? 0x02 : 0x00;
+  return kind | partiallyFillable;
+}
+
+/**
+ * A class for building a buffer of encoded orders.
+ *
+ * The encoder ensures that token addresses are kept track of and performs
+ * necessary computation in order to map each token addresses to IDs.
+ */
+export class OrderEncoder {
+  private readonly _tokens: string[] = [];
+  private readonly _tokenMap: Record<string, number | undefined> = {};
+  private _encodedOrders = "0x";
+
+  /**
+   * Creates a new order encoder instance.
+   * @param domainSeparator Optionally specified domain separator used for
+   * signing orders to encode. See {@link hashOrder} for more details.
+   */
+  public constructor(public readonly domainSeparator?: string) {}
```

Comment:
Why is the domain separator optional?

Reply:
Ah, this was so you can use the `OrderEncoder` without one, if you already have order signatures from somewhere else.

Reply:
I think it would be clearer if it weren't optional for now. We can re-evaluate later if needed.